Feature: Sort auras by name, author and update

bentsku opened the pull request:
This pull request adds the feature request in #658 

I also made small fixes and refactoring that I can delete before doing another PR.

The main implementation is the sorting of the auras list. I've used computed properties to return a sort function, depending on the column selected and the ascending/descending order. 
This sort function is then used in the computed properties which provide the data to render the list.
The sort order for the column "update ready" might not be correct, because I didn't have any dummy data to test it. 

I've made several small changes that maybe shouldn't be in the same pull request ?

1. Cleaned up a little the computed properties, many of them were doing the same thing, I've used them to calculate the next ones (_see commit db4f0fa_). It allows me to use only one computed property to render the list.

2. Also fixed a little alignment bug in the Config page, when the WoW Path was empty, the icons and fake input wouldn't be align vertically. (_commit 1ceb84a_)

3. I've encountered a network error, and I'm not sure commit c5ec85e fixes it because I didn't reproduce the network error again. I would correctly load the first API call, but the subsequent call to individual aura would timeout, and the error wouldn't be caught in the promise chain and the "Fetch" button would be there forever. 

4. Created a function that generate the defaultValues instead on relying on JSON.stringify and parse. We can call the function to get the defaultValues back. We might need to rename it in getDefaultValues, haven't done it yet. (_commit 3b0e065_)

5. Lastly, I used a template tag to group to configStep condition under one tag. It allows to use v-else-if on the 3 other pages. (_commit 46ef208_)

I hope this is to the standard of your pull requests, I'd love to participate more on this project! I've been working with Vue since 2 years now, and I appreciate being able to use it outside of work.

mrbuds commented:
This is my first (and only) vue.js project, I'm glad you are interested in helping with your experience in the field.
The code is bit messy and may not always follow best practices, a lot of refactoring can and should be done, your changes are very welcome.

I'll need a bit of time to make sure there is no regression in this PR.
Feel free to join us on discord at https://discord.gg/wa2 and type `!role alpha` in `#bot-channel` so you see `#companion-app` channel in development sub-category

cont1nuity commented:
The sorting should take `auraType` or `auraTypeDisplay` into account. For Plater, this value is filled to distinguish mods, scripts and profiles.
I can see this working fine when sorting the `Updates` column, but I think it can work as a separate sort column as well.

Example: https://i.imgur.com/l82ZJBr.png

bentsku commented:
Thanks! The code isn't that messy, I got around pretty quickly. And it is working, I think that's the main part. I'll try to dedicate some time to try things and refactor some parts, thanks again for the blessing. 

@cont1nuity, I'll work to add the other column now. I don't user Plater, and I'm missing dummy data to work on, I'm working on my Mac and WoW is on a Windows laptop. Had to copy some parts to try. I'll add a column header for `auraType` and `auraTypeDisplay`. Thanks for the feedback!

cont1nuity commented:
@bentsku you can use the attached `Plater.lua` for testing. Just drop it in your SavedVariables.
And/Or give me a ping for testing.
[Plater.zip](https://github.com/WeakAuras/WeakAuras-Companion/files/5322164/Plater.zip)

cont1nuity commented on src/components/LandingPage.vue:
```diff
@@ -120,6 +120,20 @@
                   </span>
                 </div>
+                <div
+                  v-if="addonSelected == 'Plater'"
```
I'd prefer this to be not hard-coded but possibly tied to a check if values exist in the column.
The reason for this lies in the modular approach to adding other supported addons which might use this column as well.
Possibly add a flag in the `addonConfigs` and check for the current selected addon tab?

bentsku replied:
I get it, it didn't feel right to me either. I can either check if one of the value in the column has the property auraTypeDisplay, that'd be the easiest.
But like you said, I can use the `addonConfigs`.I can either use a flag, or an array with the possibles column values, in case we add some, or want more modularity.

cont1nuity replied:
Thank you. Looks good to me.

bentsku commented:
Really sorry for the spam of commit, I'll stop for tonight. Added the new column to sort, and sorry for the typo. I've also refactored the headers into a component to help for the review, and keep the LandingPage template around the same length

cont1nuity commented:
Two things I like to suggest: 
- Secondary sort order: When sorting for "updates", it could be good to have a "secondary sort order". E.g.: sort by update, then name and last by type. This way there would be a good one-glance-overview over the updates.
- Having the header not scroll with the list would be great. That way you can sort directly, even far down in the list. Especially useful if you have "show auras without updates" enabled in the settings.

bentsku commented:
For the first thing: I see what you mean. If I understood well, you're talking about the update column, to chain the sorting for a clearer view of the update, in a predetermined way?
Maybe for the author and the type, a secondary sort would make sense too, to not follow the order it comes before any sorting.

Should these sort orders be customisable ? I could create a chain sort by clicking one column after another, but it could become a bit tricky to reset the sorts without any button.
The sorts could also be predetermined (by us) for each column, like you suggested. 
I guess author column could be sorted by name then type too. And for the type column, I guess the name as a secondary sort could be enough. I'll write some example and share it to validate them.

About the header, I didn't realised it actually! It definitely makes sense. I think I need to take the headers out of the `#aura-list`.

I'll work on that tomorrow and come back with those changes. 
Thanks for the feedback! Really appreciated.

bentsku commented:
Hello!
So, the headers are now above the scroll area. If we change the sort, should we scroll back up or just stay where we are ?

For the secondary indexes, I for now opted for fixed columns. I had to create a new utility js file, `sort.js`, to lighten my sortFunction computed property. 

I created 2 functions to sort the columns `updates` and `author`.
They follow the same logic, depending on the presence of the `auraTypeDisplay` column. 

1. If the main column (the one clicked) has identical values, we then sort on the type (if present, always A->Z) then on the name (always A->Z as well, to not flip the whole table and keep consistency).
Example: 

| Name | Update v | Type | Author | rest
| -- | -- | -- | -- | -- |
| A name  | Update ready | A Type | Author A | [...]
| B name  | Update ready  | A Type | Author D | [...]
| A name  | Update ready  | B Type | Author B | [...]
| A name  | Update ready  | C Type | Author F | [...]
| A name  | skipped update  | C Type | Author F | [...]

| Name | Update ^ | Type | Author | rest
| -- | -- | -- | -- | -- |
| A name  | -- | A Type | Author A | [...]
| B name  | --  | A Type | Author D | [...]
| A name  | skipped update  | B Type | Author B | [...]
| A name  | skipped update  | C Type | Author F | [...]
| A name  | Update ready  | C Type | Author F | [...]

The type column is sorted on itself, then on the name of aura. 
I hope this is clear.

I've been wondering about the fact that we don't use the `modified` column anymore on these sorts. I guess it makes it more readable ?

Stanzilla commented:
Hey there, sorry for the wait, I'll be testing this after the pre-patch hype dies down

cont1nuity commented:
From a first test I'd say: works like a charm. Resetting the sort will default back to "update timestamp", which is the only column without a sort header.
I can't give much feedback about the implementation, though, as I am not too good with vue/js, but it looks versatile enough be enhanced if needed.

bentsku commented:
No problem about the wait, I thought so, there must be a lot of work on WA, good luck for that!
And thanks for the feedback cont1nuity!

mrbuds commented on src/components/LandingPage.vue:
```diff
@@ -35,91 +35,86 @@
+              :auras-shown="aurasSortedForView.length"
+            ></RefreshButton>
+            <br />
+            <AuraHeaders
```
Headers should be hidden when there is no aura to show
`:class="{ hidden: aurasSortedForView.length <= 0 }"`

![image](https://user-images.githubusercontent.com/189656/95924209-718a5700-0db7-11eb-8cbb-f659f7e18047.png)

bentsku replied:
I was thinking about creating a component `AuraList` holding both `AuraHeaders` and `AuraListItem`, which would take care of the sorting and would not "pollute" the code of the `LandingPage`. The only thing is that for now, the sort parameters would be held inside of the list component, so if we change page and go on Settings or others, the sort parameters would reset. Would that be acceptable ? 
That would be until we could refactor the config object into something more global, and not held by the `LandingPage`.

mrbuds commented:
Thanks and congratulations for the merge :)

cont1nuity commented:
Thank you!

bentsku commented:
Thank you! I'll start working on the refactoring now, I guess we should translate the aura's headers also, I've just did the English ones.  But thanks again for the opportunity!
